core/app/models/spree/shipping_rate.rb

```diff
@@ -1,46 +1,56 @@
 module Spree
   class ShippingRate < Spree::Base
+    attr_writer :order
+
     belongs_to :shipment, class_name: 'Spree::Shipment'
     belongs_to :shipping_method, -> { with_deleted }, class_name: 'Spree::ShippingMethod', inverse_of: :shipping_rates
-    belongs_to :tax_rate, -> { with_deleted }, class_name: 'Spree::TaxRate'
 
-    delegate :order, :currency, to: :shipment
-    delegate :name, to: :shipping_method
+    has_many :taxes,
+             class_name: "Spree::ShippingRateTax",
+             foreign_key: "shipping_rate_id",
+             dependent: :destroy
+
+    delegate :currency, to: :shipment
+    delegate :order, to: :shipment, prefix: true
+    delegate :name, :tax_category, to: :shipping_method
     delegate :code, to: :shipping_method, prefix: true
+    alias_attribute :amount, :cost
 
-    def display_base_price
-      Spree::Money.new(cost, currency: currency)
-    end
+    extend DisplayMoney
+    money_methods :amount
+    alias_method :discounted_amount, :amount
 
-    def calculate_tax_amount
-      tax_rate.calculator.compute_shipping_rate(self)
+    def order
+      @order || shipment_order
     end
 
     def display_price
-      price = display_base_price.to_s
-      if tax_rate
-        tax_amount = calculate_tax_amount
-        if tax_amount != 0
-          if tax_rate.included_in_price?
-            if tax_amount > 0
-              amount = "#{display_tax_amount(tax_amount)} #{tax_rate.name}"
-              price += " (#{Spree.t(:incl)} #{amount})"
-            else
-              amount = "#{display_tax_amount(tax_amount * -1)} #{tax_rate.name}"
-              price += " (#{Spree.t(:excl)} #{amount})"
-            end
-          else
-            amount = "#{display_tax_amount(tax_amount)} #{tax_rate.name}"
-            price += " (+ #{amount})"
-          end
-        end
-      end
-      price
+      price = display_amount.to_s
+
+      return price if taxes.empty? || amount == 0
+
+      tax_explanations = taxes.map(&:label).join(tax_label_separator)
+
+      Spree.t :display_price_with_explanations,
+               scope: 'shipping_rate.display_price',
+               price: price,
+               explanations: tax_explanations
     end
     alias_method :display_cost, :display_price
 
-    def display_tax_amount(tax_amount)
-      Spree::Money.new(tax_amount, currency: currency)
+    def pre_tax_amount
+      default_vat_amount_sum = Spree::TaxRate.
```

Comment:
I'm having a hard time wrapping my head around part of this.  In current Solidus doesn't the pre_tax_amount depend on where the item is being shipped (i.e. depend on which tax rate should be used)?  Yet here we are always using the tax rates for the default zone right?  Doesn't that mean that the code below from default_tax.rb will compute the wrong amount if the order is actually being shipped somewhere other than the default zone?
```ruby
def compute_item(item)
  if rate.included_in_price
    deduced_total_by_rate(item.pre_tax_amount, rate)
  else
    round_to_two_places(item.discounted_amount * rate.amount)
  end
end
```
Where am I getting mixed up?